Sources/Workspace/WorkspaceConfiguration.swift

```diff
@@ -209,6 +209,118 @@ extension Workspace {
+            case .workspaceAndUser(let rootPath):
+                // package/project "local" .netrc file, takes priority over user-level file
+                let localPath = rootPath.appending(component: ".netrc")
+                // user didn't tell us to explicitly use these .netrc files so be more lenient with errors
+                if let localProvider = self.loadOptionalNetrc(path: localPath, fileSystem: fileSystem, observabilityScope: observabilityScope) {
+                    providers.append(localProvider)
+                }
+
+                // user .netrc file (most typical)
+                let userHomePath = fileSystem.homeDirectory.appending(component: ".netrc")
+                // user didn't tell us to explicitly use these .netrc files so be more lenient with errors
+                if let userHomeProvider = self.loadOptionalNetrc(path: userHomePath, fileSystem: fileSystem, observabilityScope: observabilityScope) {
+                    providers.append(userHomeProvider)
+                }
+            case .user:
+                // user .netrc file (most typical)
+                let userHomePath = fileSystem.homeDirectory.appending(component: ".netrc")
+                
+                // user didn't tell us to explicitly use these .netrc files so be more lenient with errors
+                if let userHomeProvider = self.loadOptionalNetrc(path: userHomePath, fileSystem: fileSystem, observabilityScope: observabilityScope) {
+                    providers.append(userHomeProvider)
+                }
+            case .disabled:
+                // noop
+                break
+            }
+
+            switch self.keychain {
+            case .enabled:
+                #if canImport(Security)
+                providers.append(KeychainAuthorizationProvider(observabilityScope: observabilityScope))
+                #else
+                assertionFailure("Keychain not supported on this platform")
+                #endif
+            case .disabled:
+                // noop
+                break
+            }
+
+            return providers.isEmpty ? .none : CompositeAuthorizationProvider(providers, observabilityScope: observabilityScope)
+        }
+
+        private func loadOptionalNetrc(
+            path: AbsolutePath,
+            fileSystem: FileSystem,
+            observabilityScope: ObservabilityScope
+        ) -> NetrcAuthorizationProvider? {
+            guard fileSystem.exists(path) else {
+                return .none
+            }
+
+            do {
+                // FIXME: need new API in FileSystem to test for read and write access
+                // test for read access
+                try _ = fileSystem.getDirectoryContents(path.parentDirectory)
+            } catch {
+                observabilityScope.emit(warning: "\(path.parentDirectory) is not accessible or not readable, not using .netrc file in it. \(error)")
+                return .none
```

Comment:
@yim-lee this is my version of the fix in #4075, I would like to also add APIs to FileSystem to check for readability/writability so we dont have to hack around it with temp files, etc

Reply:
reference: rdar://88308095

Reply:
We can't merge this PR without the FileSystem API then, correct? It's the writability that's problematic.

Reply:
no in this case it needs to be able to read, which this test is a good approximation for. I am planning to add the FS APIs in a separate PR as its a more involved change

Reply:
@yim-lee this is now using the new readability API

Reply:
See my comment in https://github.com/apple/swift-package-manager/pull/4079#discussion_r798045066. We need `isWritable` unless we decide to remove write capability from `NetrcAuthorizationProvider`.